### ISSUE TITLE
Incorrect generated types & fix for "eslint@typescript-eslint/unbound-method" error

### ISSUE THREAD

**Issue by IlCallo** [intlify/vue-i18n-next#324]
### vue & vue-i18n version
ex:
- vue: 3.0.5
- vue-i18n: 9.0.0-rc.4

### What is actually happening?
Hi there!
I'm testing the new package version and I got stuck for a while on a couple of problems related to the generated types:
- `declare` keyword used on `type` and `interface` exports, which has no effect (AFAIK)
- declaration of `useI18n` functions using method style (`t(): string`) instead of property style (`t: () => string`), which causes a "eslint@typescript-eslint/unbound-method" linting error when destructured

The first problem shouldn't have any side effect, it's just pointless: `declare` keyword has no effect when used in that way. You should probably remove those anyway, just to be sure.

The second one would force anyone using TypeScript linter to disable an error every time the composable is used.
I guess you defined those using the method notation to be able to add overloads, but TS assume that functions on object defined with method notation MIGHT be using the local context, that's why the linter then complains about this.
You can accomplish the same thing by declaring those functions and overloads at root level, then adding a property with the function name and with type as the `typeof` of the declared function.

Example: 
```ts
declare function t(key: Path | number): string;
declare  function t(key: Path | number, plural: number, options?: TranslateOptions): string;
declare function t(key: Path | number, defaultMsg: string, options?: TranslateOptions): string;
// Other overloads

export interface Composer<Messages = {}, DateTimeFormats = {}, NumberFormats = {}, Message = VueMessageType> {
    // other properties
    t: typeof t;
    // other properties
}
```

I tried this out on-the-fly and it seems to work correctly.
Other Vue3-ready packages I used doesn't seem to share this problem

**Comment by kazupon** [comment]
Hmm, 🤔 
The `*.d.ts` of  vue-i18n-next is generated with api-extractor.
https://github.com/microsoft/rushstack/tree/master/apps/api-extractor

I think that api-extractor have problems.

**Comment by IlCallo** [comment]
I'm not familiar with API Extractor (yet).
Dunno how you can solve the first problem, but I the second one is probably due to how you define the composer interface in https://github.com/intlify/vue-i18n-next/blob/master/packages/vue-i18n/src/composer.ts

If you extract `t` (and other functions) as root level definitions, in a similar fashion to what I suggested into the first post, it's probable that API Extractor will be able to generate typings correctly and fix the second error (which right now is the one actually causing real problems)

**Comment by kazupon** [comment]
Yeah, we can solve the second problem, but we run into the problem of not being able to generate API references automatically.

The API references in vue-i18n-next use api-extractor and api-docs-gen and generator.
For cases like `t: typeof t`, We can use `@inheritDoc` to copy from another comment.
However, it is not currently working. The issue has been registered in an issue...
https://github.com/microsoft/rushstack/issues/2062

As we are planning to release v9 officially soon, I want to avoid stopping the automatic generation of API reference.

Acturally, I also checked about `@typescript-eslint/unbound-method`, but it seems to be **warning** by default.
I prefer that you would like to use eslint-disable (e.g. `/* eslint-disable-line @typescript-eslint/unbound-method */`) to work around it.

**Comment by IlCallo** [comment]
> I prefer that you would like to use eslint-disable (e.g. /* eslint-disable-line @typescript-eslint/unbound-method */) to work around it.

We surely can, but this means everyone using this package with Composition API and TypeScript will have to disable that rule **every time** it's destructured (or disable it altogether, which isn't really acceptable). This translates to using it nearly many times as the number of your components (if you're using pure Composition API style you suggest).

It seems strange all composables from other Vue ecosystem packages don't have this problems, and some of them use api-extractor too (eg. `vue` itself).
Maybe they had the same problem at some point and found a way to workaround this?

Anyway, I understand you don't want to mess things up right now, it's more a DX problem.
Since I was exploring this to write Quasar 2 migration guide, I'll just suggest everyone with the problematic setup to stick to the legacy API (if possible) or disable the rule inline

Thanks for looking into this anyway!

**Comment by DarkLite1** [comment]
We've stumbled upon this warning too:
![image](https://user-images.githubusercontent.com/10828570/118930462-0d7bd280-b946-11eb-9a0f-cfca17a94203.png)

The funny thing is, before migrating to Vue3 we used the package `vue-I18n-composable`. Using this same code with [that package](https://github.com/intlify/vue-i18n-composable) didn't generate an error or warning from the rule `@typescript-eslint/unbound-method:`. Maybe you can have a look there on how they circumvented this?

**Comment by IlCallo** [comment]
https://github.com/intlify/vue-i18n-composable/blob/f35ee68712fcc1358a37b9d2f1632f58a06bccd8/src/index.ts#L44-L48

That's because they export the methods references as already bound to a context in a way that TS can understand

**Comment by Shayan-To** [comment]
As a workaround, you can add this to some `.ts` or `.d.ts` file that has an `import` or `export` in it (you can add `export {};`):

```
/* eslint-disable */
declare module "vue-i18n" {
    type Same<T> = { [k in keyof T]: T[k] };

    export type Composer2<
        Messages = {},
        DateTimeFormats = {},
        NumberFormats = {},
        Message = VueMessageType,
    > = Same<Composer<Messages, DateTimeFormats, NumberFormats, Message>>;

    export function useI18n<Options extends UseI18nOptions = object>(
        options?: Options,
    ): Composer2<Options["messages"], Options["datetimeFormats"], Options["numberFormats"]>;
}
/* eslint-enable */
```

**Comment by IlCallo** [comment]
This seems resolved in latest 9.2 beta version
